### ISSUE TITLE
SingletaskStratifiedSplitter Doesn't satisfy k-fold API

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#752]
Copying over from #736:

```
In [6]: splitter = dc.splits.SingletaskStratifiedSplitter()

In [7]: splitter.k_fold_split(train, 4)
TIMING: dataset construction took 7.125 s
Loading dataset from disk.
TIMING: dataset construction took 7.511 s
Loading dataset from disk.
TIMING: dataset construction took 7.562 s
Loading dataset from disk.
TIMING: dataset construction took 7.643 s
Loading dataset from disk.
Out[7]: 
[<deepchem.data.datasets.DiskDataset at 0x7f17ad480190>,
 <deepchem.data.datasets.DiskDataset at 0x7f17acf7c8d0>,
 <deepchem.data.datasets.DiskDataset at 0x7f17af477550>,
 <deepchem.data.datasets.DiskDataset at 0x7f17afcdc910>]
```

`splitter.k_fold_split()` should return a list of tuples `(train, valid)` instead of returning raw datasets.

**Comment by chf42** [comment]
@rbharath Thank you for opening a new issue for that topic. 

The following question is after I did splitter.k_fold_split, how could I get access to each subset? 

According to documents, it returns "list of length k tuples of (train, valid)". For example, k_splitter = splitter.k_fold_split(train,4). k_splitter[0] should give me (train1,valid1). k_splitter[0][0] should give me 'train1'. However, when I trained **model.fit(k_splitter[0][0])**, I have errors like "TypeError: 'NumpyDataset' object does not support indexing". However, I can run **model.fit(k_splitter[0])** successfully. I am very confused there... because I want to set up cross-validation for k-fold, I assume that I need to train 'train1' and predict 'valid1', and repeat until 'train k' and predict 'valid k'.

Need help there....

Thanks,
Cheng

**Comment by rbharath** [comment]
I recommend using a different splitter for the time being. Your proposed code should work for the new API.

**Comment by rbharath** [comment]
Marking this issue as open for contribution. To implement this feature, please modify `SingletaskStratifiedSplitter` to return a list of `(train,valid)` tuples in the style of other splitters and add a unit test that verifies that the splitter returns data in the correct format.

**Comment by VIGS25** [comment]
I was looking at the code for the k_fold_split for the SingletaskStratifiedSplitter and that for the general Splitter class and they are completely different. 

The docs mention that, for the SingletaskStratifiedSplitter, the k_fold_split from the base class was overridden. Is there a reason for this?

The same issue is also present in the RandomStratifiedSplitter.

**Comment by rbharath** [comment]
I don't believe there's a fundamental reason. I think we found it easier to override when we first implemented, but I think we could well refactor and clean-up nicely for both `SingletaskStratifiedSplitter` and `RandomStratifiedSplitter`.
